File: python/02_流程控制/02_循环语句.py
# ============================================================
# Python 流程控制 - 循环语句
# ============================================================

print("=== 1. for 循环 - 遍历列表 ===")
fruits = ["苹果", "香蕉", "橙子", "葡萄"]
for fruit in fruits:
    print(f"水果: {fruit}")

print("\n=== 2. for 循环 - range() 函数 ===")
print("range(5):")
for i in range(5):  # 0 到 4
    print(i, end=" ")
print()

print("range(2, 7):")
for i in range(2, 7):  # 2 到 6
    print(i, end=" ")
print()

print("range(1, 10, 2):")
for i in range(1, 10, 2):  # 1 到 9，步长为2
    print(i, end=" ")
print()

print("\n=== 3. for 循环 - 带索引遍历 ===")
for index, fruit in enumerate(fruits):
    print(f"索引 {index}: {fruit}")

print("\n=== 4. while 循环 ===")
count = 1
while count <= 5:
    print(f"第 {count} 次循环")
    count += 1

print("\n=== 5. break 语句（跳出循环） ===")
for i in range(10):
    if i == 5:
        break
    print(i, end=" ")
print("\n遇到 break，循环结束")

print("\n=== 6. continue 语句（跳过当前迭代） ===")
for i in range(10):
    if i % 2 == 0:
        continue
    print(i, end=" ")
print("\n只打印了奇数")

print("\n=== 7. 循环中的 else 子句 ===")
for i in range(5):
    print(i, end=" ")
else:
    print("\n循环正常结束（没有被 break 中断）")

print("\n=== 8. 嵌套循环 ===")
print("九九乘法表:")
for i in range(1, 10):
    for j in range(1, i + 1):
        print(f"{j}x{i}={i*j}\t", end="")
    print()

print("\n=== 9. 使用示例：累加求和 ===")
total = 0
for i in range(1, 101):
    total += i
print(f"1到100的和: {total}")

print("\n=== 10. 使用示例：查找质数 ===")
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n ** 0.5) + 1):
        if n % i == 0:
            return False
    return True

print("1到50之间的质数:")
for num in range(1, 51):
    if is_prime(num):
        print(num, end=" ")
print()

File: rust/02_流程控制/02_循环语句.rs
// ============================================================
// Rust 流程控制 - 循环语句
// 编译运行：rustc 02_循环语句.rs -o 02_循环语句 && ./02_循环语句
// ============================================================

fn main() {
    println!("=== 1. loop 循环 ===");

    // 基本 loop
    let mut counter = 0;
    let result = loop {
        counter += 1;
        if counter == 10 {
            break counter * 2;  // break 可以返回值
        }
    };
    println!("循环结果: {}", result);

    // 无限循环（用 break 退出）
    println!("\n--- 无限循环示例 ---");
    let mut n = 0;
    loop {
        if n >= 3 {
            break;
        }
        println!("n = {}", n);
        n += 1;
    }

    println!("\n=== 2. while 循环 ===");

    let mut number = 3;
    while number != 0 {
        println!("{}!", number);
        number -= 1;
    }
    println!("发射！");

    // while let
    println!("\n--- while let ---");
    let mut stack = vec![1, 2, 3];
    while let Some(top) = stack.pop() {
        println!("弹出: {}", top);
    }

    println!("\n=== 3. for 循环 ===");

    // 遍历数组
    let a = [10, 20, 30, 40, 50];
    for element in a.iter() {
        println!("值为: {}", element);
    }

    // Range
    println!("\n--- Range ---");
    for number in 1..4 {  // 1..4 是 1, 2, 3
        println!("{}", number);
    }

    // rev() 反转
    println!("\n--- rev() 反转 ---");
    for number in (1..4).rev() {
        println!("{}!", number);
    }
    println!("出发！");

    // 带索引遍历
    println!("\n--- enumerate() 带索引 ---");
    let fruits = ["苹果", "香蕉", "橙子"];
    for (i, fruit) in fruits.iter().enumerate() {
        println!("索引 {}: {}", i, fruit);
    }

    println!("\n=== 4. break 和 continue ===");

    print!("break 示例（到5停止）: ");
    for i in 0..10 {
        if i == 5 {
            break;
        }
        print!("{} ", i);
    }
    println!();

    print!("continue 示例（跳过偶数）: ");
    for i in 0..10 {
        if i % 2 == 0 {
            continue;
        }
        print!("{} ", i);
    }
    println!();

    println!("\n=== 5. 循环标签 ===");

    // 使用循环标签跳出多层循环
    let mut count = 0;
    'outer: loop {
        println!("外层循环: {}", count);
        let mut inner = 0;
        loop {
            println!("  内层循环: {}", inner);
            if inner == 3 {
                break 'outer;  // 跳出外层循环
            }
            inner += 1;
        }
    }
    println!("跳出所有循环");

    println!("\n=== 6. 使用示例：查找质数 ===");
    println!("1 到 50 之间的质数:");
    for num in 1..=50 {
        if is_prime(num) {
            print!("{} ", num);
        }
    }
    println!();

    println!("\n=== 7. 使用示例：斐波那契数列 ===");
    println!("前 10 项:");
    for i in 0..10 {
        print!("{} ", fibonacci(i));
    }
    println!();

    println!("\n=== 8. 使用示例：乘法表 ===");
    for i in 1..=5 {
        for j in 1..=i {
            print!("{}*{}={:<2} ", j, i, i * j);
        }
        println!();
    }

    println!("\n=== 运行完成 ===");
}

fn is_prime(n: i32) -> bool {
    if n < 2 {
        return false;
    }
    for i in 2.. {
        if i * i > n {
            break;
        }
        if n % i == 0 {
            return false;
        }
    }
    true
}

fn fibonacci(n: u32) -> u32 {
    if n <= 1 {
        return n;
    }
    let mut a = 0;
    let mut b = 1;
    for _ in 2..=n {
        let temp = a + b;
        a = b;
        b = temp;
    }
    b
}
